### PR TITLE
use multiple grpc connections for one tikv store

### PR DESCRIPTION
**Pull request by hhkbp2**
Hi,

This PR uses multiple grpc connections for one tikv store.

PTAL @disksing @tiancaiamao @coocood

### CODE REVIEW

**Comment by hhkbp2** [comment]
PTAL @disksing @tiancaiamao @coocood

File: store/tikv/client.go [hunk 1]
```diff
@@ -48,6 +49,58 @@ type Client interface {
...
+		}
+		a.v[a.index] = conn
+	}
+	a.index = (a.index + 1) % a.maxSize
+	return conn, nil
+}
+
+func (a *connArray) Close() {
+	a.lock.Lock()
+	for i := uint32(0); i < a.maxSize; i++ {
```
Comment (disksing) [inline store/tikv/client.go]: for i, conn in range a.v

File: store/tikv/client.go [hunk 2]
```diff
@@ -48,6 +49,58 @@ type Client interface {
...
+	maxSize uint32
+	index   uint32
+	v       []*grpc.ClientConn
+}
+
+func newConnArray(maxSize uint32, addr string) *connArray {
+	return &connArray{
+		lock:    &sync.Mutex{},
+		addr:    addr,
+		maxSize: maxSize,
```
Comment (disksing) [inline store/tikv/client.go]: I think it's redundant to store _maxSize_, you can use _len(a.v)_.

File: store/tikv/client.go [hunk 3]
```diff
@@ -48,6 +49,67 @@ type Client interface {
...
+		v:     make([]*grpc.ClientConn, maxSize),
+	}
+	if err := a.Init(maxSize, addr); err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+func (a *connArray) Init(maxSize uint32, addr string) error {
+	for i, _ := range a.v {
```
Comment (disksing) [inline store/tikv/client.go]: `for i := range a.v`
The rest LGTM.

File: store/tikv/client.go [hunk 4]
```diff
@@ -48,6 +49,67 @@ type Client interface {
...
+		index: 0,
+		v:     make([]*grpc.ClientConn, maxSize),
+	}
+	if err := a.Init(maxSize, addr); err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+func (a *connArray) Init(maxSize uint32, addr string) error {
```
Comment (tiancaiamao) [inline store/tikv/client.go]: The maxSize argument is redundant.

**Comment by hhkbp2** [comment]
PTAL @tiancaiamao @coocood @siddontang

File: store/tikv/client.go [hunk 5]
```diff
@@ -48,6 +49,67 @@ type Client interface {
...
+		a.v[i] = conn
+	}
+	return nil
+}
+
+func (a *connArray) Get() *grpc.ClientConn {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	conn := a.v[a.index]
+	a.index = (a.index + 1) % uint32(len(a.v))
```
Comment (siddontang) [inline store/tikv/client.go]: I prefer atomic add here

Comment (siddontang) [inline store/tikv/client.go]: do we need to check conn is nil after Close?

**Comment by tiancaiamao** [comment]
LGTM

File: store/tikv/client.go (hunk 5 above)
Comment (hhkbp2) [inline store/tikv/client.go]: It's not necessary since the whole connection map in `rpcClient` is protected by `rpcClient.isClosed`.

**Pull request merged**